src/datasets.cpp

```diff
@@ -520,23 +512,43 @@ DataFrame DataFrame::sample(int nrow, int seed) const{
     }
     // number of rows to pull
     if (nrow == -1){
+        // if default, bootstrap a sample of equal length
         nrow = this->length();
+    } else if (replace == false){
+        // if sample without replacing, cap at original length
+        nrow = std::min(nrow, this->length());
     }else{
         assert(nrow > 0);
     }
-    // Seed the generator
-    std::mt19937 eng(seed);
-    // Draw row indices from uniform distribution
-    std::uniform_int_distribution<> distr(0, this->length()-1);
     // Create new empty DataFrame
     DataFrame new_frame = DataFrame();
-    // pull random rows (as pointers, not copies) until full
-    while (new_frame.length() < nrow)
-    {
-        // get random row index with replacement
-        int rand_row = distr(eng);
-        // get pointer to that row in original dataframe and store in bootstrap
-        new_frame.addRow(this->row(rand_row));
+    if (replace == true){
+        // Seed the generator
+        std::mt19937 eng(seed);
+        // Draw row indices from uniform distribution
+        std::uniform_int_distribution<> distr(0, this->length()-1);
+        // pull random rows (as pointers, not copies) with replacement until full
+        while (new_frame.length() < nrow){
+            // get random row index with replacement
+            int rand_row = distr(eng);
+            // get pointer to that row in original dataframe and store in bootstrap
+            new_frame.addRow(this->row(rand_row));
+        }
+    }else{
+        // pre-allocate vector of row indices
+        std::vector<int> indices(this->length());
+        // fill vector, equivalent to np.arange(0, this->length()-1)
+        std::generate(indices.begin(), indices.end(), [n = 0] () mutable {return n++;});
```

Comment:
Per [this SO](https://stackoverflow.com/questions/50798713/c-generating-random-vector), you may be able to use distr() to directly generate random numbers when filling indices. This might save time/code, but current implementation works.

Reply:
If you mean the `distr` in the preceding `if` statement, I avoided using that as it just samples from the same uniform distribution each time, meaning it's effectively using replacement. In the highlighted code it shuffles the indices so it can pop them one at a time without replacing.

That said, I might full well be misunderstanding you. In either case, I've no deep affection for the code, so if you have a better suggestion (which seems probable) I'm more than open to improvements! 😀